### PR TITLE
add description for key update.

### PR DESCRIPTION
**Pull request by jyao1**
Fix https://github.com/riscv-non-isa/riscv-ap-tee-io/issues/55
Fix https://github.com/riscv-non-isa/riscv-ap-tee-io/issues/53

### CODE REVIEW

File: specification/07-theory_operations.adoc [hunk 1]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
...
+session keys before the use of keys reach to the limit.
+
+==== SPDM Session Key Update
+
+SPDM session key update is managed by SPDM KEY_UPDATE command.
+The TSM shall count the usage of SPDM session keys, also knowns as the 64bit
+`sequence number` defined in <<SecuredSPDM>>.
+Whenever the VMM asks the TSM to encrypt the SPDM message, the TSM shall check
+if the `sequence number` reaches to the limit. If yes, the TSM shall notify VMM to
+perform SPDM Key Update action (through the `sbi_covh_spdm_key_update()` `COVH` call).
```
Comment (sameo) [inline specification/07-theory_operations.adoc:893]: The VMM never explictly asks the TSM to encrypt SPDM messages, but sends `COVH` requests that could lead to the generation and encryption of SPDM messages from the TSM. See https://github.com/riscv-non-isa/riscv-ap-tee-io/blob/main/specification/07-theory_operations.adoc#spdm-transport

So I suggest this instead:

```suggestion
Whenever the TSM generates and encrypts SPDM messages to support the host supervisor domain manager `COVH` requests, it shall check if the `sequence number` reaches to the limit. If it does, the TSM notifies the VMM to initiate an SPDM Key Update (through the `sbi_covh_spdm_key_update()` `COVH` call), by replying with an `SBI` error to the corresponding `COVH` request.
Once the SPDM Key Update completes, the VMM shall re-send the failing `COVH` request. 
```

File: specification/07-theory_operations.adoc [hunk 2]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
...
+
+==== IDE Session Key Refresh
+
+IDE session key refresh is managed by host software via IDE_KM command.
+The host root port IDE engine shall count the 64bit invocation field of the IV,
```
Comment (sameo) [inline specification/07-theory_operations.adoc:898]:
```suggestion
The host root port IDE engine shall count the 64-bit invocation field of the IV,
```

File: specification/07-theory_operations.adoc [hunk 3]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
...
+and check if the value reaches to the limit, then it process the IDE TLP.
```
Comment (sameo) [inline specification/07-theory_operations.adoc:899]:
```suggestion
check if the value reaches to the limit, and then process the IDE TLP.
```

File: specification/07-theory_operations.adoc [hunk 4]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
...
+If overflow happens, the IDE stream shall be in insecure state.
```
Comment (sameo) [inline specification/07-theory_operations.adoc:900]: Is it a `must move to insecure state` here as well?

File: specification/07-theory_operations.adoc [hunk 5]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
...
+In order to keep the IDE session alive, the VMM should perform IDE Key Refresh
+action (through the `sbi_covh_ide_key_refresh()` `COVH` call) before the IDE
+session reaches the limit of the key usage.
+
+Currently, there is no robust way to get the invocation field of IV for a given
+IDE session. One possible way is to do time-based refresh. For example, the VMM
+perform the IDE key refresh every several minutes based on the calculation of
```
Comment (sameo) [inline specification/07-theory_operations.adoc:907]:
```suggestion
could periodically perform IDE key refresh, based on the calculation of
```

File: specification/07-theory_operations.adoc (hunk 4 above)
Comment (sameo) [inline specification/07-theory_operations.adoc:900]:
```suggestion
If the invocation field overflows, the IDE stream moves to the insecure state and the IDE session terminates.
```

File: specification/09-coveio_abi.adoc [hunk 6]
```diff
@@ -270,6 +270,47 @@ The possible error codes returned in `sbiret.error` are shown below.
 | SBI_ERR_FAILED          | The operation failed for unknown reasons.
 |===
 
+[#sbi_covh_spdm_key_update]
+===== Function: CoVE Host SPDM Key Update (FID #1033)
+[source, C]
+-----
+struct sbiret sbi_covh_spdm_key_update()(unsigned long device_id);
```
Comment (sameo) [inline specification/09-coveio_abi.adoc:277]:
```suggestion
struct sbiret sbi_covh_spdm_key_update(unsigned long device_id);
```

File: specification/09-coveio_abi.adoc [hunk 7]
```diff
@@ -270,6 +270,47 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+| Error code              | Description
+| SBI_SUCCESS             | The operation completed successfully.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+[#sbi_covh_ide_key_refresh]
+===== Function: CoVE Host IDE Key Refresh (FID #1034)
+[source, C]
+-----
+struct sbiret sbi_covh_ide_key_refresh()(unsigned long device_id,
```
Comment (sameo) [inline specification/09-coveio_abi.adoc:297]:
```suggestion
struct sbiret sbi_covh_ide_key_refresh(unsigned long device_id,
```

File: specification/09-coveio_abi.adoc [hunk 8]
```diff
@@ -270,6 +270,47 @@ The possible error codes returned in `sbiret.error` are shown below.
...
 |===
 
+[#sbi_covh_spdm_key_update]
+===== Function: CoVE Host SPDM Key Update (FID #1033)
+[source, C]
+-----
+struct sbiret sbi_covh_spdm_key_update()(unsigned long device_id);
+-----
+
+Perform SPDM key update for a device.
```
Comment (sameo) [inline specification/09-coveio_abi.adoc:280]:
```suggestion 
Perform an SPDM key update for a the SPDM connection between the TSM and a connected device.
```

File: specification/09-coveio_abi.adoc [hunk 9]
```diff
@@ -270,6 +270,47 @@ The possible error codes returned in `sbiret.error` are shown below.
...
+
+[#sbi_covh_ide_key_refresh]
+===== Function: CoVE Host IDE Key Refresh (FID #1034)
+[source, C]
+-----
+struct sbiret sbi_covh_ide_key_refresh()(unsigned long device_id,
+                                         unsigned long stream_id);
+-----
+
+Perform IDE key refresh for a stream on a device.
```
Comment (sameo) [inline specification/09-coveio_abi.adoc:301]:
```suggestion
Perform an IDE key refresh for a connected device IDE stream.
```

File: specification/07-theory_operations.adoc [hunk 10]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
 
 === Interface Unbinding
 
+=== Session Key Update
+
+In TEE-IO architecture, there are 2 sessions established between the host TSM
+and the device: SPDM session and IDE session. Both sessions use AES-GCM-256
+keys to protect the data. As such, the host software needs to update the
```
[the remark this answers is not in the document]
Comment (jyao1) [reply specification/07-theory_operations.adoc:883]: I don't think we can put mandatory requirement for an untrusted software.

File: specification/07-theory_operations.adoc [hunk 11]
```diff
@@ -876,6 +876,37 @@ TVM ->> TVM: Use device interface
 
 === Interface Unbinding
 
+=== Session Key Update
+
+In TEE-IO architecture, there are 2 sessions established between the host TSM
+and the device: SPDM session and IDE session. Both sessions use AES-GCM-256
```
[the remark this answers is not in the document]
Comment (jyao1) [reply specification/07-theory_operations.adoc]: done

File: specification/07-theory_operations.adoc (hunk 1 above)
Comment (jyao1) [reply specification/07-theory_operations.adoc]: done

File: specification/07-theory_operations.adoc (hunk 3 above)
Comment (jyao1) [reply specification/07-theory_operations.adoc]: done

File: specification/07-theory_operations.adoc (hunk 4 above)
Comment (jyao1) [reply specification/07-theory_operations.adoc]: done

File: specification/09-coveio_abi.adoc (hunk 6 above)
Comment (jyao1) [reply specification/09-coveio_abi.adoc]: done

File: specification/09-coveio_abi.adoc (hunk 7 above)
Comment (jyao1) [reply specification/09-coveio_abi.adoc]: done

File: specification/09-coveio_abi.adoc (hunk 8 above)
Comment (jyao1) [reply specification/09-coveio_abi.adoc]: done

File: specification/09-coveio_abi.adoc (hunk 9 above)
Comment (jyao1) [reply specification/09-coveio_abi.adoc]: done

File: specification/07-theory_operations.adoc (hunk 2 above)
Comment (jyao1) [reply specification/07-theory_operations.adoc]: done

**Review by sameo — APPROVED** [review]

File: specification/07-theory_operations.adoc [hunk 12]
```diff
@@ -876,6 +876,47 @@ TVM ->> TVM: Use device interface
...
+`sbi_covh_spdm_key_update()` `COVH` call).
+
+Once the SPDM Key Update completes, the VMM shall re-send the failing `COVH`
+request.
+
+==== IDE Session Key Refresh
+
+IDE session key refresh is managed by host software via IDE_KM command.
+The host root port IDE engine shall count the 64-bit invocation field of the IV,
+check if the value reaches to the limit, and then process the IDE TLP.
```
[the remark this answers is not in the document]
Comment (jyao1) [reply specification/07-theory_operations.adoc]: done

**Pull request merged**